pkg/acs/calacs/acsccd/blev_funcs_postsm4.c

```diff
@@ -36,152 +39,218 @@ static int make_amp_array(const int arr_rows, const int arr_cols, SingleGroup *
   const double dsi_sens[NAMPS] = {2.9188919e-3, 10.805754e-3, 12.432145e-3, 3.8596253e-3};
-    
-  /* factor combining time constant and clocking frequency */
-  double factor;
-  
-  /* summation variables */
-  double sum;
-  int num;
-  double magic_square_mean;
-  
-  /* allocate space for data arrays */
-  ampdata = malloc(arr_rows * arr_cols * sizeof(double));
-  ampdata_gap = malloc(nquad_pix * sizeof(double));
-  
-  /* allocate space for true DC bias levels array */
+
+  PtrRegister ptrReg;
+  initPtrRegister(&ptrReg);
+
+  /* array of true DC bias levels */
+  double * dc_bias_levels = NULL;
   dc_bias_levels = malloc((nquad_pix + 1) * sizeof(double));
-  
-  for (i = 0; i < NAMPS; i++) {
-    /* put a single amp's data into ampdata */
-    if (i < 2) {
-      make_amp_array(arr_rows, arr_cols, chip1, i, ampdata);
-    } else {
-      make_amp_array(arr_rows, arr_cols, chip2, i, ampdata);
-    }
-       
-    /* calculate "magic square mean" */
-    sum = 0.0;
-    num = 0;
+  addPtr(&ptrReg, dc_bias_levels, &free);
     
-    for (j = 2057; j <= 2066; j++) {
-      for (k = 13; k <= 22; k++) {
-        sum += ampdata[arr_cols*j + k];
-        num++;
+  /* array for amp data and amp data + gap pixels */
+  double *ampdata = malloc(arr_rows * arr_cols * sizeof(double));
+  addPtr(&ptrReg, ampdata, &free);
+  double *ampdata_gap = malloc(nquad_pix * sizeof(double));
+  addPtr(&ptrReg, ampdata_gap, &free);
+  
+  /* For full frame data, there are four amps in use.  For supported subarray
+   * data, there is only one amp in use.  The actual amp in use must be determined. 
+   */
+  int numAmpsInUse = (acs->subarray == YES) ? 1 : NAMPS;
+  
+  /* Loop over all the amps for full frame data (total of 4 amps, 0 to 3) or just use 
+   * the single amp value corresponding to the supported subarray (1 amp, value defined in acs.h).
+   */
+  int ampInUse = 0;
+  {unsigned int i;
+  for (i = 0; i < numAmpsInUse; i++) {
+      ampInUse = i;
+      if ((i == 0) && (acs->subarray == YES)) {
+         if (strcmp (acs->ccdamp, "A") == 0)
+            ampInUse = AMP_A;
+         else if (strcmp (acs->ccdamp, "B") == 0)
+            ampInUse = AMP_B;
+         else if (strcmp (acs->ccdamp, "C") == 0)
+            ampInUse = AMP_C;
+         else if (strcmp (acs->ccdamp, "D") == 0)
+            ampInUse = AMP_D;
+      } 
+
+      /* put a single amp's data into ampdata */
+      if (acs->subarray == YES) {
+         make_amp_array(arr_rows, arr_cols, sg[0], ampInUse, ampdata);
+      } else {
+         if (i < 2) {
+            make_amp_array(arr_rows, arr_cols, sg[1], ampInUse, ampdata);
+         } else {
+            make_amp_array(arr_rows, arr_cols, sg[0], ampInUse, ampdata);
+         }
       }
-    }
+       
+      /* calculate "magic square mean" - this works for any single amp */
+      double sum = 0.0;
+      int num = 0;
+      double magic_square_mean = 0.0;
     
-    magic_square_mean = sum / (double) num;
+      {unsigned int j, k;
```

Comment:
This would be better as a function, in which case the hardcoded boundaries may be less of an issue.

Reply:
Is there a way to check that the amp being passed in even has this region? Should the check be in ``bias_shift_corr()``, I would have perhaps thought so.